unit/java_bytecode/java_types/generic_type_index.cpp

```diff
@@ -0,0 +1,65 @@
+/*******************************************************************\
+
+ Module: Unit tests for java_types
+
+ Author: Diffblue Ltd.
+
+\*******************************************************************/
+
+#include <testing-utils/catch.hpp>
+#include <java_bytecode/java_types.h>
+
+SCENARIO("generic_type_index", "[core][java_types]")
+{
+  GIVEN("Generic type LGenericClass<TX;Tvalue;>; and parameters X, value, Z")
+  {
+    const auto symbol_type = symbol_typet("java::GenericClass");
+    const auto generic_symbol_type = java_generic_symbol_typet(
+      symbol_type, "LGenericClass<TX;Tvalue;>;", "PrefixClassName");
+    java_generic_parametert paramX("PrefixClassName::X", symbol_typet());
+    java_generic_parametert value("PrefixClassName::value", symbol_typet());
+    java_generic_parametert paramZ("PrefixClassName::Z", symbol_typet());
+
+    WHEN("Looking for parameter indexes")
+    {
+      const auto indexX = generic_symbol_type.generic_type_index(paramX);
+      const auto index_value = generic_symbol_type.generic_type_index(value);
+      const auto indexZ = generic_symbol_type.generic_type_index(paramZ);
+
+      THEN("X has index 0, Y index 1 and Z is not found")
+      {
+        REQUIRE(indexX.has_value());
+        REQUIRE(indexX.value() == 0);
+        REQUIRE(index_value.has_value());
+        REQUIRE(index_value.value() == 1);
+        REQUIRE_FALSE(indexZ.has_value());
+      }
+    }
+  }
+
+  GIVEN("Generic type Ljava/util/HashMap<TK;TV;>; and parameters K, V, T")
```

Comment:
Sorry for being nit picking here but as I thought about the application of `java_generic_symbol_typet` I would suggest to change these examples a bit to reflect why that type even exists. It is to represent the generic base of a class. So you could change the examples to say that for example given `HashMap<K,V> extends Map<K,V>` the generic symbol for Map is created correctly - the generic symbol is then created with `java_generic_symbol_typet(symbol_typet("java::java.util.Map"), "Ljava/util/Map<TK;TV;>;", "java.util.HashMap")` (exactly as in the documentation for the constructor actually). Similarly for the first example - `PrefixClassName<X,value> extends GenericClass<X,value>` and and the constructor call stays as it is.